### ISSUE TITLE
JSF 2.3.15: Websocket trying to reconnect infinitely instead of finitely (because of browser specific bug)

### ISSUE THREAD

**Issue by Toru47** [eclipse-ee4j/mojarra#4933]
As said in https://github.com/eclipse-ee4j/mojarra/issues/4572#issuecomment-822392768 the websocket bug "Websocket trying to reconnect infinitely instead of finitely (because of browser specific bug)" is obviously not fixed in version 2.3.15. I've tried this version in WildFly 21.0.0.Final added 2.3.15 Mojarra manually with the same error result. This error occurs if the application is (hot) redeployed in Eclipse:

![websocket](https://user-images.githubusercontent.com/47194731/118804320-d9040a00-b8a4-11eb-83ec-98cda6391fdd.png)
![websocket2](https://user-images.githubusercontent.com/47194731/118804963-a575af80-b8a5-11eb-9b69-c3438986a231.png)

Test with Eclipse 4.18, WildFly 21.0.0.Final with manually added JSF 2.3.15 Mojarra, Firefox 62.0.3 or MS Edge 90.0.818

Bean:

```java
@Named
@SessionScoped
public class FooBean implements Serializable {
    private static final long serialVersionUID = -9057373496036307133L;
    @Inject
    @Push
    private PushContext fooChannel;

    public void test() {
        fooChannel.send("Test");
    }
}
```
Facelet:
```html
<f:websocket channel="fooChannel" onmessage="function(m){ ... }" scope="session" />
```
It doesn't matter if the bean is ViewScoped or SessionScoped.

**Comment by BalusC** [comment]
Reproduced. This is after all not a problem in client side, let alone a "browser specific bug". It happens in every single browser. This is after all actually a problem in server side. The server log shows an infinite loop of
```
java.lang.NullPointerException
	at com.sun.faces.cdi.CdiUtils.getBeanReferenceByType(CdiUtils.java:221)
	at com.sun.faces.cdi.CdiUtils.getBeanReference(CdiUtils.java:204)
	at com.sun.faces.push.WebsocketSessionManager.getInstance(WebsocketSessionManager.java:268)
	at com.sun.faces.push.WebsocketEndpoint.onOpen(WebsocketEndpoint.java:65)
```
The technical problem is, it was expecting to find the CDI bean manager in FacesContext, but there is no FacesContext anywhere at the moment the websocket hits the server.

Have you checked the server log?

It worked in OmniFaces because it uses `CDI.current().getBeanManager()` but this was nowhere used in Mojarra.

**Comment by Toru47** [comment]
I haven't got any errors in server log, not even in debug mode.

Now I've build this fixed version (Mojarra 2.3.16-SNAPSHOT) with

`impl/src/main/java/com/sun/faces/util/Util.java (98189bea9d485c8bf459718eae1e127b063c4fad)`

and run in WildFly 21

`08:46:10,346 INFO  [javax.enterprise.resource.webcontainer.jsf.config] (ServerService Thread Pool -- 81) Mojarra 2.3.16-SNAPSHOT...`

but the issue is not fixed. After doing a

`[standalone@localhost:9990 /] reload`

and reload the page the server floods the browser with the requests mentioned above - infinitely long. The browser consumes a lot of CPU. If I close the browser and open it again with that page, the problem still exists. So, closing and opening the the browser is not a solution to stop that flooding. Only the stopping of the server will stop that flooding.

Moreover the problem seems to be browser and session independent. If I open another browser along with the first one that error occurs in the second browser too.
